src/transformers/modeling_rag.py

```diff
@@ -0,0 +1,950 @@
+        encoder_outputs (:obj:`tuple(tuple(torch.FloatTensor)`, `optional`, defaults to :obj:`None`):
+            Tuple consists of (`last_hidden_state`, `optional`: `hidden_states`, `optional`: `attentions`, `doc_scores`)
+            `last_hidden_state` of shape :obj:`(batch_size, n_docs * sequence_length, hidden_size)` is a sequence of hidden-states at the output of the last layer of the encoder.
+            `doc_scores` of shape :obj:`(batch_size, n_docs)` store retrieval scores of documents retrieved for each input in the batch.
+            Used by the (:class:`~transformers.RagToken`) model during decoding.
+        decoder_input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, target_sequence_length)`, `optional`, defaults to :obj:`None`):
+            Provide for generation tasks. `None` by default, constuct as per instructions for the generator model you're using with your RAG instance.
+        decoder_past_key_values (:obj:`tuple(tuple(torch.FloatTensor))`):
+            Tuple consists of two elements: ``encoder_outputs`` of the RAG model (see ``encoder_outputs``) and ``decoder_past_key_values`` of the underlying generator.
+            Can be used to speed up decoding. ``decoder_past_key_values`` are used in the (:class:`~transformers.RagToken`)
+            model during decoding.
+        use_cache (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            If `use_cache` is True, ``decoder_past_key_values`` are returned and can be used to speed up decoding (see
+            ``decoder_past_key_values``).
+        generator_kwargs (remaining dictionary of keyword arguments, `optional`):
+            Additional keyword arguments will be passed to the generator forward pass.
+"""
+
+RAG_LOSS_INPUTS_DOCSTRING = r"""
+        return_loss (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            If :obj:`True`, computes the loss which is returned as part of the :class:`~transformers.file_utils.Seq2SeqLMOutputWithDocs`.
+            Otherwise, loss defaults to :obj:`None`.
+        reduce (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Only relevant if ``return_loss`` is set to :obj:`True`. If :obj:`True`, the NLL loss is reduced using the ``torch.Tensor.sum`` operation.
+        label_smoothing (:obj:`float`, `optional`, defaults to ``0.0``):
+            Only relevant if ``return_loss`` is set to :obj:`True`. Controls the ``epsilon`` parameter value for label smoothing in the loss calculation.
+            If set to ``0.0``, no label smoothing is performed.
+"""
+
+
+@add_start_docstrings_to_callable(RAG_CORE_DOCSTRING)
+class RagModel(torch.nn.Module):
+    def __init__(
+        self,
+        config,
+        retriever,
+        question_encoder,
+        question_encoder_tokenizer,
+        generator,
+        generator_tokenizer,
+    ):
+        super().__init__()
+        self.config = config
+        self.retriever = retriever
```

Comment:
Do you think the retriever could be outside the model ?
I think it could could make things easier for people who want to use a different way to retrieve documents, or a different distributed setup.

I noticed that calling `contextualize` is the first step of forward and generate but maybe it could be called outside those functions ? I guess `contextualize` would be called inside `_step` or `_generative_step` instead.